### PR TITLE
add more supports for multiple groups; change the version of experimental api v1alpha1

### PR DESCRIPTION
**Pull request by caesarxuchao**
Breaking #12755 into smaller pieces. This is the first PR, it includes:
1. 6df2870: enforces the URL exposed by apiserver to be in the form of prefix/group/version/...
   - for api/v1, we set the APIPrefix="api", group="", version="v1", so the URL is still /api/v1
   - for experimental api, we set the APIPrefix="api", group="experimental", version="v1alpha1", so the URL is /api/experimental/v1alpha1.
   - for openshift v1 API, because it's exposed at /oapi/v1 now, we can either make prefix="oapi", group="", or prefix="", group="oapi", so the URL will not be changed.
2. 9a139b6: stop apiserver from encoding api.Status to a specific version unless the version is v1.
3. the commit "refactor pkg/api/testapi to support multiple groups. Unit tests fixed." implements #13233.

There will be subsequent PR depending on this one:
1. factoring out default version, oldest version, storage version, etc. https://github.com/kubernetes/kubernetes/pull/12755#discussion_r37247918
2. add KubeAPIPrefix https://github.com/kubernetes/kubernetes/pull/12755#discussion_r37251228
3. move api.Status to unversioned https://github.com/kubernetes/kubernetes/pull/12755#discussion_r37251228
4. move /expapi to /api/experimental https://github.com/kubernetes/kubernetes/pull/12755#discussion_r37251963
5. ease our testing for multiple groups/versions
   https://github.com/kubernetes/kubernetes/pull/12755#discussion_r37329030

@uluyol @nikhiljindal @bgrant0607 @krousey @smarterclayton @derekwaynecarr

### CODE REVIEW

**Comment by smarterclayton** [comment]
Can we summarize the current status of the proposal into a proposal document or design (if one doesn't exist)?  I have a lot of concerns about this but I'm finding I can't keep up with the pace of the design.  I think this is a fundamental part of the core product we're changing, so if anything deserves a proposal it's this.  I'd also like to offer my feedback on the pieces that are important to us (as an external API group that has different needs than expapi) and I'm worried about spreading around these pulls.

**Comment by smarterclayton** [comment]
@bgrant0607 re my comment about this needing a proposal

**Comment by caesarxuchao** [comment]
@smarterclayton, I will write one.

**Comment by caesarxuchao** [comment]
cc @lavalamp, this is the first step of #12591.

**Comment by caesarxuchao** [comment]
@nikhiljindal, you can ignore the second commit "don't encode api.Status to a version if the version is not v1", because we changed our plan for it.

File: pkg/api/util/group_version.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import "strings"
+
+func GetVersion(groupVersion string) string {
```
Comment (nikhiljindal) [inline pkg/api/util/group_version.go]: How about combining these 2 methods into a single getGroupAndVersion().
Then you can return an error if version != "v1" and group is empty.

File: pkg/api/util/group_version.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (nikhiljindal) [inline pkg/api/util/group_version.go]: 2015

File: pkg/api/util/group_version.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+package util
+
+import "strings"
+
+func GetVersion(groupVersion string) string {
+	s := strings.Split(groupVersion, "/")
+	if len(s) != 2 {
+		//e.g. return "v1" for groupVersion="v1"
```
Comment (nikhiljindal) [inline pkg/api/util/group_version.go]: Add a space after "//"

File: cmd/genconversion/conversion.go [hunk 4]
```diff
@@ -71,12 +72,19 @@ func main() {
 	}
 
 	versionPath := path.Join(pkgBase, group, version)
+	if group == "experimental" {
+		//TODO: we should rename the direcotry /expapi to /experimental, so directory path match groupVerson
```
Comment (nikhiljindal) [inline cmd/genconversion/conversion.go]: type in directory

File: cmd/genconversion/conversion.go [hunk 5]
```diff
@@ -71,12 +72,19 @@ func main() {
...
+		versionPath = path.Join(pkgBase, "expapi", version)
+	}
 	generator := pkg_runtime.NewConversionGenerator(api.Scheme.Raw(), versionPath)
 	apiShort := generator.AddImport(path.Join(pkgBase, "api"))
 	generator.AddImport(path.Join(pkgBase, "api/resource"))
 	// TODO(wojtek-t): Change the overwrites to a flag.
 	generator.OverwritePackage(version, "")
-	for _, knownType := range api.Scheme.KnownTypes(version) {
+	if *groupVersion == "api/v1" {
```
Comment (nikhiljindal) [inline cmd/genconversion/conversion.go]: Is this to not break existing clients who might be passing api/v1 explicitly?
If yes, then shouldnt we be doing this at the top? Maybe just after flag.Parse.
Also add a comment saying this to support old clients, who might still be sending api/v1, and we can delete it once we stop supporting v1

File: cmd/gendeepcopy/deep_copy.go [hunk 6]
```diff
@@ -63,13 +64,16 @@ func main() {
...
-	group, version := path.Split(*groupVersion)
-	group = strings.TrimRight(group, "/")
+	group := apiutil.GetGroup(*groupVersion)
+	version := apiutil.GetVersion(*groupVersion)
 	registerTo := "api.Scheme"
 	if *groupVersion == "api/" {
 		registerTo = "Scheme"
 	}
 	pkgname := group
+	if group == "experimental" {
```
Comment (nikhiljindal) [inline cmd/gendeepcopy/deep_copy.go]: How about keeping "experimental" and "expapi" as constants in apiutils?

File: cmd/gendeepcopy/deep_copy.go [hunk 7]
```diff
@@ -92,7 +100,14 @@ func main() {
 			generator.OverwritePackage(vals[0], vals[1])
 		}
 	}
-	for _, knownType := range api.Scheme.KnownTypes(version) {
+	//handle the special cases of groupVersion = "api/" and "experimental/"
+	if version == "" {
```
Comment (nikhiljindal) [inline cmd/gendeepcopy/deep_copy.go]: Shouldnt we be doing this only if group == "api"?
For ex: for groupVersion="users/", we do want to pass "users/" to api.Scheme.KnownTypes() right?

Comment (nikhiljindal) [inline cmd/gendeepcopy/deep_copy.go]: I think we can fix this by adding:
```
if groupVersion == "api/v1" || groupVersion =="api/" {
  groupVersion = groupVersion[4:len(groupVersion)-1]
}
```
at the top.

File: cmd/kube-apiserver/app/server.go [hunk 8]
```diff
@@ -332,7 +329,7 @@ func (s *APIServer) Run(_ []string) error {
 
 	// "experimental/v1={true|false} allows users to enable/disable the experimental API.
```
Comment (nikhiljindal) [inline cmd/kube-apiserver/app/server.go]: update the comment as well

File: hack/after-build/update-generated-conversions.sh [hunk 9]
```diff
@@ -39,13 +39,23 @@ EOF
...
-	mv "$TMPFILE" "pkg/${version}/conversion_generated.go"
+    #TODO: remove this case if we rename the directory expapi/ to experimental/
+    if [ "${version%/*}" == "experimental" ]; then
+        mv "$TMPFILE" "pkg/expapi/${version##*/}/conversion_generated.go"
+    else
+	    mv "$TMPFILE" "pkg/${version}/conversion_generated.go"
+    fi
 }
 
-DEFAULT_VERSIONS="api/v1 expapi/v1"
```
Comment (nikhiljindal) [inline hack/after-build/update-generated-conversions.sh]: replace api/v1 by v1

File: hack/after-build/update-generated-conversions.sh [hunk 10]
```diff
@@ -39,13 +39,23 @@ EOF
 
   "${genconversion}" -v "${version}" -f - >>  "$TMPFILE"
 
-	mv "$TMPFILE" "pkg/${version}/conversion_generated.go"
+    #TODO: remove this case if we rename the directory expapi/ to experimental/
```
Comment (nikhiljindal) [inline hack/after-build/update-generated-conversions.sh]: When you do that, I will recommend moving it to pkg/api/experimental. All groups in future like users, admin, etc can come under pkg/api (pkg/api/admin, pkg/api/users, pkg/api/experimental).

File: hack/after-build/update-generated-deep-copies.sh [hunk 11]
```diff
@@ -56,11 +60,15 @@ function generate_deep_copies() {
...
 		# KUBE_API_VERSIONS.
-		apiVersions="${ver##*/}"
-		KUBE_API_VERSIONS="${apiVersions}" generate_version "${ver}"
-	done
+		if [[ ${ver##*/} == "v1" ]] || [[ ${ver##*/} == "" ]]; then
+			apiVersions="${ver##*/}"
+		else 
+			apiVersions="${ver}"
+		fi
+			KUBE_API_VERSIONS="${apiVersions}" generate_version "${ver}"
```
Comment (nikhiljindal) [inline hack/after-build/update-generated-deep-copies.sh]: wrong indentation?

File: hack/test-go.sh [hunk 12]
```diff
@@ -266,7 +267,7 @@ reportCoverageToCoveralls() {
 }
 
 # Convert the CSVs to arrays.
-IFS=',' read -a apiVersions <<< "${KUBE_TEST_API_VERSIONS}"
+IFS=';' read -a apiVersions <<< "${KUBE_TEST_API_VERSIONS}"
```
Comment (nikhiljindal) [inline hack/test-go.sh]: Is this intended?

File: hack/test-go.sh [hunk 13]
```diff
@@ -274,7 +275,7 @@ for (( i=0, j=0; ; )); do
   apiVersion=${apiVersions[i]}
   etcdPrefix=${etcdPrefixes[j]}
   echo "Running tests for APIVersion: $apiVersion with etcdPrefix: $etcdPrefix"
-  KUBE_API_VERSION="${apiVersion}" KUBE_API_VERSIONS="v1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
+  KUBE_TEST_API="${apiVersion}" KUBE_API_VERSIONS="v1,experimental/v1alpha1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
```
Comment (nikhiljindal) [inline hack/test-go.sh]: All tests pass for experimental/v1?

File: hack/test-integration.sh [hunk 14]
```diff
@@ -43,24 +43,27 @@ runTests() {
...
     KUBE_RACE="-race" \
     KUBE_TEST_API_VERSIONS="$1" \
-    KUBE_API_VERSIONS="v1" \
+    KUBE_API_VERSIONS="v1,experimental/v1alpha1" \
     "${KUBE_ROOT}/hack/test-go.sh" test/integration
 
   kube::log::status "Running integration test scenario"
 
- KUBE_API_VERSIONS="v1" "${KUBE_OUTPUT_HOSTBIN}/integration" --v=${LOG_LEVEL} --api-version="$1" \
+ # TODO: fix the integration test to make the --api-version meaningful for integration test
```
Comment (nikhiljindal) [inline hack/test-integration.sh]: Is this TODO still valid?

File: hack/test-integration.sh [hunk 15]
```diff
@@ -43,24 +43,27 @@ runTests() {
...
+ # KUBE_API_VERSIONS="v1,experimental/v1alpha1" "${KUBE_OUTPUT_HOSTBIN}/integration" --v=${LOG_LEVEL} --api-version="$1" \
+ #  --max-concurrency="${KUBE_INTEGRATION_TEST_MAX_CONCURRENCY}"
+ KUBE_API_VERSIONS="v1,experimental/v1alpha1" "${KUBE_OUTPUT_HOSTBIN}/integration" --v=${LOG_LEVEL} --api-version="v1" \
   --max-concurrency="${KUBE_INTEGRATION_TEST_MAX_CONCURRENCY}"
 
   cleanup
 }
 
-KUBE_API_VERSIONS="v1" "${KUBE_ROOT}/hack/build-go.sh" "$@" cmd/integration
+KUBE_API_VERSIONS="v1,experimental/v1alpha1" "${KUBE_ROOT}/hack/build-go.sh" "$@" cmd/integration
```
Comment (nikhiljindal) [inline hack/test-integration.sh]: I expect integrations test to fail atleast since we do not expose experimental by default

File: hack/test-integration.sh [hunk 16]
```diff
@@ -43,24 +43,27 @@ runTests() {
...
 
 # Run cleanup to stop etcd on interrupt or other kill signal.
 trap cleanup EXIT
 
 # Convert the CSV to an array of API versions to test
-IFS=',' read -a apiVersions <<< "${KUBE_TEST_API_VERSIONS}"
+IFS=';' read -a apiVersions <<< "${KUBE_TEST_API_VERSIONS}"
```
Comment (nikhiljindal) [inline hack/test-integration.sh]: Why this change?

File: pkg/api/serialization_test.go [hunk 17]
```diff
@@ -86,14 +86,22 @@ func roundTrip(t *testing.T, codec runtime.Codec, item runtime.Object) {
...
 // roundTripSame verifies the same source object is tested in all API versions.
-func roundTripSame(t *testing.T, item runtime.Object, except ...string) {
+func roundTripSame(t *testing.T, item runtime.Object, kind string, except ...string) {
 	set := util.NewStringSet(except...)
 	seed := rand.Int63()
 	fuzzInternalObject(t, "", item, seed)
+
 	version := testapi.Version()
+	codec := testapi.Codec()
+	//TODO: we should deduct if a kind belongs to "experimental" group once we
```
Comment (nikhiljindal) [inline pkg/api/serialization_test.go]: Did you mean "detect" instead of "deduct"?

**Comment by nikhiljindal** [comment]
I just glanced over the later part of PR as its a big one.
I have added some comments, which should give you enough to work on for now.

**Comment by caesarxuchao** [comment]
#12837 enables validate by default and it reveals the problem that validation doesn't work across groups of different versions. I'm fixing the problem.

**Comment by caesarxuchao** [comment]
client.ServerAPIVersions() needs to be fixed. Currently it reports:
Server does not support API version 'experimental/v1alpha1'. Falling back to 'v1'.

File: cmd/genconversion/conversion.go (hunk 5 above)
Comment (caesarxuchao) [inline cmd/genconversion/conversion.go]: Does other clients also generate conversion function? When I wrote this, I was only concerned with how we format the versions in hack/update-generated-conversion.sh. This seems to be the most neat way.

File: cmd/gendeepcopy/deep_copy.go (hunk 6 above)
Comment (caesarxuchao) [inline cmd/gendeepcopy/deep_copy.go]: sure. how about making that change in another PR?

File: cmd/gendeepcopy/deep_copy.go (hunk 7 above)
Comment (caesarxuchao) [inline cmd/gendeepcopy/deep_copy.go]: Sorry for the confusion. Actually the behavior here is correct, i.e., if groupVersion="users/", we want to pass "" to api.Scheme.KnownTypes(), because we want to generate deep copies for internal object, which is registered using AddKnownTypes("", ...)

File: cmd/genconversion/conversion.go (hunk 4 above)
Comment (caesarxuchao) [inline cmd/genconversion/conversion.go]: Will fix.

File: cmd/kube-apiserver/app/server.go (hunk 8 above)
Comment (caesarxuchao) [inline cmd/kube-apiserver/app/server.go]: Will fix.

File: hack/after-build/update-generated-conversions.sh (hunk 10 above)
Comment (caesarxuchao) [inline hack/after-build/update-generated-conversions.sh]: I agree. That will be another PR. Thanks.

File: hack/after-build/update-generated-conversions.sh (hunk 9 above)
Comment (caesarxuchao) [inline hack/after-build/update-generated-conversions.sh]: "api/v1" is correct here.

Sorry it looks confusing, it's because we want the code here to be consistent with the code for generating deep copies, where using "api/v1" is sort of inevitable.

File: hack/after-build/update-generated-deep-copies.sh (hunk 11 above)
Comment (caesarxuchao) [inline hack/after-build/update-generated-deep-copies.sh]: Will fix.

File: hack/test-go.sh (hunk 12 above)
Comment (caesarxuchao) [inline hack/test-go.sh]: Yes, as we discussed yesterday :)

File: hack/test-go.sh (hunk 13 above)
Comment (caesarxuchao) [inline hack/test-go.sh]: Yes.

File: hack/test-integration.sh (hunk 14 above)
Comment (caesarxuchao) [inline hack/test-integration.sh]: Yes. We need to fix integration test to support multiple groups.

File: hack/test-integration.sh (hunk 15 above)
Comment (caesarxuchao) [inline hack/test-integration.sh]: I will check again. It passed before my recent rebase.

**Comment by nikhiljindal** [comment]
Is it possible to take the test changes out of this PR?

File: pkg/api/serialization_test.go (hunk 17 above)
Comment (caesarxuchao) [inline pkg/api/serialization_test.go]: Will fix.

File: pkg/api/util/group_version.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/api/util/group_version.go]: Will fix.

File: pkg/api/util/group_version.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/api/util/group_version.go]: I'd rather not, because I want it to return single return value so that it can be used inline to generate input parameters for other functions.

File: pkg/api/util/group_version.go (hunk 3 above)
Comment (caesarxuchao) [inline pkg/api/util/group_version.go]: Will fix.

**Comment by caesarxuchao** [comment]
Unless we have a way to disable the tests related to experimental objects, they will fail without the changes to the tests.

File: hack/after-build/update-generated-deep-copies.sh [hunk 18]
```diff
@@ -27,7 +27,11 @@ gendeepcopy=$(kube::util::find-binary "gendeepcopy")
 
 function result_file_name() {
 	local version=$1
-	echo "pkg/${version}/deep_copy_generated.go"
+    if [ "${version%/*}" == "experimental" ]; then
+        echo "pkg/expapi/${version##*/}/deep_copy_generated.go"
+    else
+	    echo "pkg/${version}/deep_copy_generated.go"
```
Comment (lavalamp) [inline hack/after-build/update-generated-deep-copies.sh]: fix indentation... also tabs vs spaces

Comment (caesarxuchao) [inline hack/after-build/update-generated-deep-copies.sh]: Will fix.

**Comment by bgrant0607** [comment]
Is this PR still active?

**Comment by caesarxuchao** [comment]
No. Thanks for the reminding.

Superseded by #14156, #13833, and #13581 etc.
